gym_robotic_arm/gym_robotic_arm/env_GAIL/gail_gym_env.py

```diff
@@ -0,0 +1,112 @@
+from abc import abstractmethod
+
+import os
+import gym
+from camera import Camera
+from gym import spaces
+import numpy as np
+
+from gym_robotic_arm.constants import ARMS_LENGTHS, TOTAL_ARM_LENGTH
+
+from gym_robotic_arm.dynamic_model import RobotArm3dof
+
+
+class GailRobotArm(gym.Env):
+    # need to copy code from https://github.com/TianhongDai/hindsight-experience-replay
+    metadata = {'render.modes': ['human', 'rgb_array']}
+
+    def __init__(self, robot_controller):
+        self.robot_controller = robot_controller
+
+        # initiate camera
+        self.camera = Camera(1)
+        self.camera.show_feed_continuous()
+        
+        # camera resolution to initiate observation space
+        # laptop cam resolution and gripper/clamp cam resolution
+        self.res_l = [self.camera.laptop_cam_res[0], self.camera.laptop_cam_res[1]]
+        self.res_c = [self.camera.clamp_cam_res[0], self.camera.clamp_cam_res[1]]
+
+        # initiate obseration space
+        # code from https://github.com/openai/gym/blob/master/gym/envs/box2d/car_racing.py
+        self.observation_space = spaces.Box(
+            low=0, high=255, shape=((self.res_c[0], self.res_c[1], 3), 
+            (self.res_l[0], self.res_l[1], 3)), dtype=np.uint8
+        )
+
+        # define actions space
+        # action: {1: base joints, 2: 1st hanging joint, 3: second hanging joint,
+        #          4: gripper rotation , 5: gripper close, 6: stepper rotation}
+        self.action_space = spaces.Box(low=-1,high=1,shape=(6,),dtype=np.float32)
```

Comment:
I think it would be better to not have the action space cover all joints. Only x,z,gripper_pos for now